bwl: update the get_radio_info to include additional flags

itayx opened the pull request:
Currently all channels that are not supported get filtered out.
The get_radio_info filters flags that indicate the channel cannot be
used as an AP. the flag NO_IR is filtered as part of this process, which
is incorrect seeing as radios with radar detection can use those
channels.

Lessen the filtering check in get_radio_info for channels.
Add a "no_ir" flag to the channel_info struct

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

tomereli reviewed (approved):
Looking good. @mariomaz should review this as well.

kantera800 commented on common/beerocks/bwl/shared/nl80211_client_impl.cpp:
```diff
@@ -317,9 +317,9 @@ bool nl80211_client_impl::get_radio_info(const std::string &interface_name, radi
+                    // Ignore "channels" that are disabled or with no frequency.
                     if (!tb_freq[NL80211_FREQUENCY_ATTR_FREQ] ||
-                        tb_freq[NL80211_FREQUENCY_ATTR_DISABLED] ||
-                        tb_freq[__NL80211_FREQUENCY_ATTR_NO_IBSS]) {
```
if removing  `__NL80211_FREQUENCY_ATTR_NO_IBSS`   is part of 

> Lessen the filtering

then you should mention why in the commit message.

itayx replied:
the __NL80211_FREQUENCY_ATTR_NO_IBSS is deprecated by NL80211_FREQUENCY_ATTR_NO_IR
and we need the NO_IR flag, meaning we cannot filter it out

Updated the commit message to better reflect that

kantera800 commented on common/beerocks/bwl/shared/nl80211_client_impl.cpp:
```diff
@@ -335,6 +335,10 @@ bool nl80211_client_impl::get_radio_info(const std::string &interface_name, radi
 
+                    if (tb_freq[NL80211_FREQUENCY_ATTR_NO_IR]) {
+                        channel.no_ir = true;
+                    }
```
In the future we would want to check if radar detection is supported and if not these NO_IR (DFS) channels should not be included in the AP capabilities.

This brings me to the next point, 
I suggest that we  add `is_radar_detection_supported`  flag to the AP class , IMO should be read driver but maybe for now it can be read from prplmesh_db (need to add it there also) @arnout  what do you think
Then add this  NO_IR channels check :

```
                    // Ignore DFS channels in case AP doesn't support radar detection.
                    if (channel.no_ir && is_radar_detection_supported) {
                        continue;
                    }
```

itayx replied:
I believe a better check is to check the `is_dfs` tag as well as the `no_ir` flag

mariomaz commented on common/beerocks/bwl/shared/nl80211_client_impl.cpp:
```diff
@@ -335,6 +335,10 @@ bool nl80211_client_impl::get_radio_info(const std::string &interface_name, radi
                         channel.is_dfs = true;
                     }
 
+                    if (tb_freq[NL80211_FREQUENCY_ATTR_NO_IR]) {
```
At some point in time NL80211_FREQUENCY_ATTR_NO_IBSS and NL80211_FREQUENCY_ATTR_PASSIVE_SCAN were replaced by NL80211_FREQUENCY_ATTR_NO_IR. 

Also NL80211_FREQUENCY_ATTR_NO_IBSS was renamed to __NL80211_FREQUENCY_ATTR_NO_IBSS.

From then on, both NL80211_FREQUENCY_ATTR_NO_IR and __NL80211_FREQUENCY_ATTR_NO_IBSS
are set if NO_IR. 

However, in old kernels __NL80211_FREQUENCY_ATTR_NO_IBSS will be present but NL80211_FREQUENCY_ATTR_NO_IR will not.

You should probably do something similar to what the `iw` command line tool is doing:
```c
	/* If both flags are set assume an new kernel */
	if (tb_freq[NL80211_FREQUENCY_ATTR_NO_IR] && tb_freq[__NL80211_FREQUENCY_ATTR_NO_IBSS]) {
		print_flag("no IR", &open);
	} else if (tb_freq[NL80211_FREQUENCY_ATTR_PASSIVE_SCAN]) {
		print_flag("passive scan", &open);
	} else if (tb_freq[__NL80211_FREQUENCY_ATTR_NO_IBSS]){
		print_flag("no ibss", &open);
	}
```

itayx replied:
Changed the if statement and added a comment to describe it

mariomaz replied:
In old kernels __NL80211_FREQUENCY_ATTR_NO_IBSS will be present and NL80211_FREQUENCY_ATTR_NO_IR will not. 

We should decide on how these kernels should be treated. Shall we ignore the channel? Shall we set no_ir to true? Shall we add a no_ibss flag?

arnout replied:
I guess we should then make it a 4-state enum like the iw tool does: NO_IR, PASSIVE_SCAN, NO_IBSS or nothing at all.

Or, we could make a very simplifying assumption, and say:
- if RADAR is set, we mark it as a DFS channel;
- else if NO_IBSS is set, we ignore the channel.
That more or less corresponds to what hostapd does.

By the way, please note:
```
#define NL80211_FREQUENCY_ATTR_PASSIVE_SCAN	NL80211_FREQUENCY_ATTR_NO_IR
#define NL80211_FREQUENCY_ATTR_NO_IBSS		NL80211_FREQUENCY_ATTR_NO_IR
```
so actually the passive scan case is NO_IR without NO_IBSS. So I assume the split between NO_IR and NO_IBSS was made to support the passive scan case.

mariomaz replied:
> if NO_IBSS is set, we ignore the channel.

This is what we had before this PR. The problem is that we have some devices that set the NO_IBSS attribute and others that don't when AFAIK they should behave similarly.

itayx replied:
Added a filtering check before the end of the iteration
if the channel is not a DFS channel (denoted by the `is_dfs` flag) but still has the NO_IR or NO_IBSS attributes (the `no_ir` is set to `true`) then the channel will be ignored

mariomaz reviewed (changes requested):
This condition is not being considered (maybe because we don't have to deal with old kernels)
```c
if (!tb_freq[NL80211_FREQUENCY_ATTR_NO_IR] && tb_freq[__NL80211_FREQUENCY_ATTR_NO_IBSS]) {
    // TODO
}

arnout commented:
Note that (at least in its current state) this PR doesn't fully solve #1154. It does add the `no_ir` marker, but then that value is completely ignored. We should still remove the `no_ir` channels that are not radar channels.

kantera800 commented on common/beerocks/bwl/shared/nl80211_client_impl.cpp:
```diff
@@ -335,6 +335,13 @@ bool nl80211_client_impl::get_radio_info(const std::string &interface_name, radi
                     }
 
+                    // NL80211_FREQUENCY_ATTR_NO_IBSS and NL80211_FREQUENCY_ATTR_NO_IR are set to
+                    // NO_IR. for legacy, check __NL80211_FREQUENCY_ATTR_NO_IBSS
```
If its legacy then  why not 
```
                    if (tb_freq[NL80211_FREQUENCY_ATTR_NO_IR] ||
                        tb_freq[__NL80211_FREQUENCY_ATTR_NO_IBSS]) {
```

mariomaz replied:
That's fine as long as you don't want to distinguish older kernels and want to treat older NO_IBSS meaning as the new NO_IR.
This solution however does not consider NL80211_FREQUENCY_ATTR_PASSIVE_SCAN. I don't mean we should do it. I don't even know what all these flags mean.

kantera800 commented:
> Note that (at least in its current state) this PR doesn't fully solve #1154. It does add the `no_ir` marker, but then that value is completely ignored. We should still remove the `no_ir` channels that are not radar channels.

If by radar channels you mean dfs than I agree (pretty sure they are all dfs thought) still you cannot  use no_ir data without assuming AP has radar detection.  my suggestion is to add a flag to the ap class

itayx commented:
The `NL80211_FREQUENCY_ATTR_NO_IR` attribute is a deprecation of `NL80211_FREQUENCY_ATTR_PASSIVE_SCAN` & `NL80211_FREQUENCY_ATTR_NO_IBSS` as seen here:
```
#define NL80211_FREQUENCY_ATTR_PASSIVE_SCAN	NL80211_FREQUENCY_ATTR_NO_IR
#define NL80211_FREQUENCY_ATTR_NO_IBSS		NL80211_FREQUENCY_ATTR_NO_IR
```
So if we want to handle the different flags with an enum, that will support both the new and old version of kernel as suggested by @arnout & @mariomaz, then I suggest the following:
```
if (tb_freq[NL80211_FREQUENCY_ATTR_NO_IR]) {
    if (tb_freq[NL80211_FREQUENCY_ATTR_NO_IBSS]) {
        channel.flag = state_enum::no_ir;
    } else {
        channel.flag = state_enum::passive_scan;
    }
} else {
    if (tb_freq[NL80211_FREQUENCY_ATTR_NO_IBSS]) {
        channel.flag = state_enum::no_ibss;
    }
}
channel.flag = state_enum::NA;
```

I've also noticed an attribution flag `NL80211_FREQUENCY_ATTR_RADAR`
Its description in `linux/nl80211.h` from what I saw is
> Radar detection is mandatory on this channel in current regulatory domain.

We store this value as 
```
if (tb_freq[NL80211_FREQUENCY_ATTR_RADAR]) {
    channel.is_dfs = true;
}
```
So maybe a validation check can be something along the lines of:
```
if (channel.flag == state_enum::no_ir && !channel.is_dfs) {
    // do stuff
}
```

vitalybu reviewed (commented):
Why do we need the `passive_mode` flag anyway?
Looking at the latest `iw` [source code](https://git.kernel.org/pub/scm/linux/kernel/git/jberg/iw.git/tree/event.c#n35), it looks like the only "interesting" flags are `no_ir` and `no_ibss`.

>  NL80211_FREQUENCY_ATTR_NO_IR: no mechanisms that initiate radiation
> are permitted on this channel, this includes sending probe
> requests, or modes of operation that require beaconing.
> NL80211_FREQUENCY_ATTR_NO_IBSS: IBSS networks are not permitted
> on this channel in current regulatory domain.

So if as @arnout stated in #1154 -
> for now, we assume that hostapd and driver do support DFS. If we ever run into a situation where this isn't the case, we'll handle it at that time. It's easy to notice it because hostapd will give an error when we try to select a DFS channel.

Shouldn't we simply ignore the `no_ir` flag (or include any `no_ir` channels in the list of supported channels)?

kantera800 commented:
> 
> Shouldn't we simply ignore the `no_ir` flag (or include any `no_ir` channels in the list of supported channels)?

IMO we can't ignore these flags and just add all channels to the supported list.
DCS task works with the supported channel list so we could end up scanning a frequency that is not allowed.

mariomaz reviewed (approved):
Approving as I cannot add any more value to this discussion.

itayx committed:
bwl: update the get_radio_info to include additional flags

The get_radio_info gatherers radio information including supported
channels. these channels get filtered out if they don't match the
criteria of being able to be used as an AP for channel selection.
Currently DFS channels sometimes get filtered out

Since the __NL80211_FREQUENCY_ATTR_NO_IBSS is deprecated by
NL80211_FREQUENCY_ATTR_NO_IR the NO_IR flag is filtered as part of this
process, which is incorrect seeing as radios with radar detection can
use still those channels.

Remove the filtering of __NL80211_FREQUENCY_ATTR_NO_IBSS.
Add a no_ir flag to the channel_info struct.
Set the no_ir flag if the NL80211_FREQUENCY_ATTR_NO_IR is present.
Filter channels that are not DFS but still have the no_ir flag.

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

kantera800 reviewed (approved):
looks good.
Just fix the clang issue

itayx committed:
Merge branch 'master' into bugfix/add-additional-flags-to-channel-info
